ast/src/parsed/mod.rs

```diff
@@ -332,24 +332,119 @@ impl<Expr> Children<Expr> for SelectedExpressions<Expr> {
     }
 }
 
-#[derive(Debug, PartialEq, Eq, PartialOrd, Ord, Clone, Serialize, Deserialize, JsonSchema)]
+#[derive(Debug, Eq, PartialOrd, Ord, Clone, Serialize, Deserialize, JsonSchema)]
 pub enum Expression<Ref = NamespacedPolynomialReference> {
-    Reference(Ref),
-    PublicReference(String),
+    Reference(SourceRef, Ref),
+    PublicReference(SourceRef, String),
     // A number literal and its type.
-    Number(Number),
-    String(String),
-    Tuple(Vec<Self>),
-    LambdaExpression(LambdaExpression<Self>),
-    ArrayLiteral(ArrayLiteral<Self>),
-    UnaryOperation(UnaryOperation<Self>),
-    BinaryOperation(BinaryOperation<Self>),
-    IndexAccess(IndexAccess<Self>),
-    FunctionCall(FunctionCall<Self>),
-    FreeInput(Box<Self>),
-    MatchExpression(MatchExpression<Self>),
-    IfExpression(IfExpression<Self>),
-    BlockExpression(BlockExpression<Self>),
+    Number(SourceRef, Number),
+    String(SourceRef, String),
+    Tuple(SourceRef, Vec<Self>),
+    LambdaExpression(SourceRef, LambdaExpression<Self>),
+    ArrayLiteral(SourceRef, ArrayLiteral<Self>),
+    UnaryOperation(SourceRef, UnaryOperation<Self>),
+    BinaryOperation(SourceRef, BinaryOperation<Self>),
+    IndexAccess(SourceRef, IndexAccess<Self>),
+    FunctionCall(SourceRef, FunctionCall<Self>),
+    FreeInput(SourceRef, Box<Self>),
+    MatchExpression(SourceRef, MatchExpression<Self>),
+    IfExpression(SourceRef, IfExpression<Self>),
+    BlockExpression(SourceRef, BlockExpression<Self>),
+}
+
+// Comparison function for expressions that ignore source information.
+impl<Ref: PartialEq> PartialEq for Expression<Ref> {
```

Comment:
Do you know where this is needed?

Reply:
I remember this particular case:

https://github.com/powdr-labs/powdr/blob/e2310b1b66a4c85fddee5d8d954a170f42e8b4b8/linker/src/lib.rs#L39

But also many matches against numbers started to fail in tests.